src/Compilers/Core/Portable/Symbols/ISymbolExtensions_IsAccessibleWithin.cs

```diff
@@ -0,0 +1,348 @@
+            {
+                Debug.Assert(type.IsDefinition);
+                Debug.Assert(baseType.IsDefinition);
+
+                for (var current = type; current != null; current = current.BaseType?.OriginalDefinition)
+                {
+                    if (sameOriginalNamedType(baseType, current as INamedTypeSymbol))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            bool sameOriginalNamedType(INamedTypeSymbol t1, INamedTypeSymbol t2)
+            {
+                Debug.Assert(t1 != null);
+                Debug.Assert(t1.IsDefinition);
+                Debug.Assert(t2?.IsDefinition != false);
+
+                // We expect a given named type definition to satisfy reference identity.
+                // But we relax this to permit a type to be represented by multiple symbols (e.g. separate compilations).
+                // Note that the same symbol is expected to have the identical name as itself, despite VB language rules.
+                return t1 == t2 || t2 != null && t1.Name == t2.Name && t1.Arity == t2.Arity && sameOriginalSymbol(t1.ContainingSymbol, t2.ContainingSymbol);
+
+                bool sameOriginalSymbol(ISymbol s1, ISymbol s2)
+                {
+                    if (s1 == s2)
+                    {
+                        return true;
+                    }
+
+                    if (s1 == null || s2 == null)
+                    {
+                        return false;
+                    }
+
+                    switch (s1.Kind)
+                    {
+                        case SymbolKind.NamedType:
+                            return sameOriginalNamedType(s1 as INamedTypeSymbol, s2 as INamedTypeSymbol);
+                        case SymbolKind.Namespace:
+                            return sameNamespace(s1 as INamespaceSymbol, s2 as INamespaceSymbol);
+                        case SymbolKind.Assembly:
+                            return sameAssembly(s1 as IAssemblySymbol, s2 as IAssemblySymbol);
+                        case SymbolKind.NetModule:
+                            return sameModule(s1 as IModuleSymbol, s2 as IModuleSymbol);
+                        default:
+                            return false;
+                    }
+                }
+
+                bool sameModule(IModuleSymbol m1, IModuleSymbol m2)
+                {
+                    Debug.Assert(m1 != null);
+                    // We don't need to check the module name, as modules are effectively merged in the containing assembly.
+                    return m2 != null && sameOriginalSymbol(m1.ContainingSymbol, m2.ContainingSymbol);
+                }
+
+                bool sameAssembly(IAssemblySymbol a1, IAssemblySymbol a2)
+                {
+                    Debug.Assert(a1 != null);
+                    return a2 != null && a1.Identity.Equals(a2.Identity);
```

Comment:
>a1.Identity.Equals(a2.Identity) [](start = 41, length = 31)

I don't think this condition is appropriate to check for containment, but it appears this is what it is used for.

Reply:
It is being used to determine if the two assembly symbols identify the same assembly.

Reply:
> This is intended to align with the CLR's rules for identifying assemblies.

Please quote the section of the spec, the relevant wording used and provide explanation how it is applicable to detecting whether two type references are referencing the same definition.

---
In reply to: [185164217](https://github.com/dotnet/roslyn/pull/26367#discussion_r185164217) [](ancestors = 185164217,185124383)

Reply:
> It is being used to determine if the two assembly symbols identify the same assembly.

This is wrong, identity cannot answer this question.

---
In reply to: [185161761](https://github.com/dotnet/roslyn/pull/26367#discussion_r185161761) [](ancestors = 185161761)